teps/0015-implicit-mapping.md

```diff
@@ -0,0 +1,658 @@
+n = PipelineTasks[i].Params.size()
+k = PipelineTasks.size()
+END INITIALIZATION
+
+FOR EVERY param IN spec.params
+  specParams[param.Name] = param.Value
+END
+
+FOR i = 0; i < k; i++
+
+  IF PipelineTasks[i].taskSpec == nil THEN
+    Resolve PipelineTasks[i].taskRef to find implicit task parameters and results
+  ENDIF
+
+  Store implicit task parameters with default values and results in implicitTaskParams and taskResults[] variables respectively
+
+  FOR j = 0; j < n; j++
+    explicitTaskParams.insert(PipelineTasks[i].Params[j])
+  END
+
+  FOR j = 0; j < m; j++
+    results[taskResults[j]] = PipelineTasks[i].Name
+  END
+
+  FOR taskParam, taskDefaultValue in implicitTaskParams
+    IF explicitTaskParams[taskParam]
+      continue;
+    ENDIF
+    taskName = results[taskParams]
+    paramValue = specParam[taskParams]
+    IF taskName THEN
+      PipelineTasks[i].Params = append(PipelineTasks[i].Params, Params{taskName.result})
+    ELSEIF paramValue THEN
+      PipelineTasks[i].Params = append(PipelineTasks[i].Params, Params{paramValue})
+    ELSEIF taskDefaultValue == nil
+      return error //Param can not be matched and does not have a task default value
+    ENDIF
+  END
+
+  explicitTaskParams.erase()
+
+END
+```
+
+#### Implicit resource mapping
+
+In accordance with requirement 2, the mutating webhook resolves the `PipelineTask.Resources` field as follows:
+
+```
+INITIALIZATION
+hashmap specResources[resource.Name][resource.Value]
+hashset explicitTaskResources[resource.Name]
+implicitTaskResources[]
+k = PipelineTasks.size()
+m = PipelineTasks[i].Resources.size()
+n = implicitTaskResources.size()
+END INITIALIZATION
+
+FOR EVERY resource IN spec.resources
+  specResources[resource.Name]= resource.Value
+END
+
+FOR i = 0; i < k; i++
+
+  IF PipelineTasks[i].taskSpec == nil THEN
+    Resolve PipelineTasks[i].taskRef
+  ENDIF
+
+  Store implicit task resources in implicitTaskResources[] 
+
+  FOR j = 0; j < n; j++
+    explicitTaskResources.insert(PipelineTask[i].Resources[j])
+  END
+
+  FOR j = 0; j < n; i++;
+    IF explicitTaskResources[PipelineTask[i].Resources[j]]
+      continue
+    ENDIF
+    resourceValue = specResources[implicitTaskResources[j]]
+    IF resourceValue THEN
+      PipelineTasks[i].Resources = append(PipelineTasks[i].Resources, Resource{resourceValue})
+    ELSE
+      return error
+    ENDIF
+  ENDIF
+  END
+
+  explicitResourceParams.erase()
+
+END
+```
+
+### Alternatives
+- If a *taskRef* is used, the empty *taskSpec* field could be populated with the fetched *taskRef*.
+Instead of iterating over an implicit parameter or resource hashmap/array, Tekton can iterate over the parameters and resources in the *taskSpec*.
+Additionally, since the webhook has updated the *taskSpec* field, the reconciler would not need to fetch the *taskRef* on every reconcile loop. 
+This could be considered an attractive option that would require a separate PR.
+Partially because this can happen without implicit mapping as well.
```

Comment:
This would *most likely* be a problem on the validation side (at least compared to the current behavior). Applying the same pipeline definition (`kubectl apply -f mypipeline.yaml`) will result in an error because both `taskRef` and `taskSpec` are specified.
- If we remove that validation, which one of `taskSpec` or `taskRef` has precedence ?
- Do we overidde the `taskSpec` with the content of the Task referenced ? But then, the object is not what the user sent, and will _always_ be updated on apply.
- Do we take `taskSpec` and ignore `taskRef` ? But then, if the Task referenced changed, it will be ignored.